## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Import Data

In [2]:
# Folder path

path = r'C:\Users\Windows 10\Documents\04-2023 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Windows 10\\Documents\\04-2023 Instacart Basket Analysis'

In [4]:
# Import products.csv

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import orders_wrangled.csv

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Consistency Check

### Mixed-Type Data

In [6]:
# Create a new test dataframe
# The first command, df_test = pd.DataFrame(), creates a new dataframe called df_test. 

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
# The second command, df_test['mix'] = ['a', 'b', 1, True], creates a new column, mix, within df_test and fills it with numeric, string, and boolean values.

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

# The structure that’s being used in this code is called a “for-loop.”
# The “for” in for-loop stands for “for these elements, do this,”
# and the “loop” describes how the structure works: looping over and over again as it performs the procedures detailed by the “for.”
# Here, the for-loop is looping through each column in the dataframe and executing the same block of code each time.    
    
# Within the for-loop, a new variable is created: weird.
# Assigned to it is a test that checks whether the data types within the column are consistent. 
# The weird variable will ultimately take a boolean value of either True or False. 
# If True, that means the column contains inconsistent data types. 
# If False, that means the column contains only one data type.
# Boolean values can also be represented by numbers: 0 as False and 1 as True.

# Here comes the “if” statement.
# An if statement checks if some condition is met, and if it’s met, executes a line of code.
# If the condition isn’t met, the code isn’t executed.
# Here, the if statement is checking whether weird is true or false.
# If it’s greater than 0, than it’s true. If not, it’s false.
# If weird is true, the command print(col) is executed, which prints the problematic column for you to see.
# Because of the for-loop, this command will be executed on every column in your dataframe, printing every mixed-type column it finds.

mix


In [11]:
# Changing data type to string

df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# Recheck for mixed data type

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')


mix  consistent


### Missing Values

In [13]:
df_prods.isnull().sum()

# What this does is assign the function isnull() to the df_prods dataframe, then sum the result with the attached sum() function.
# The isnull() function is used to find missing observations, with “observations” here referring to entries in your dataframe.
# Think of them like cells in Excel.

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# To actually view these 16 values, you can create a subset of the dataframe

df_nan = df_prods[df_prods["product_name"].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values

In [16]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [17]:
df_prods.median()

C:\Users\Windows 10\AppData\Local\Temp\ipykernel_10232\3387230063.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [ ]:
# Replacing missing value with mean
# df['column with missings'].fillna(mean value, inplace=True)

In [ ]:
# Replacing missing value with median
# df['column with missings'].fillna(median value, inplace=True)

In [ ]:
# Looking at the rows in df_nan, it quickly becomes clear that there isn’t much you can do in terms of imputation.
# The missing values here are product names, which are strings.String values can’t be imputed like numeric values.
# You can either remove the missing values entirely or
# filter out the ones that aren’t missing into a subset dataframe and continue your analysis with this new dataframe.

#### Records with missing product names will be filtered out and a new dataframe created

In [18]:
#  The shape function returns the number of rows and columns in a dataframe.

df_prods.shape

(49693, 5)

In [19]:
# Create dataframe without missing values in 'product_name' column

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
# Compare row and column numbers for new dataframe
# Your new dataframe should have exactly 16 less rows than the original dataframe (the same as the number of missing values).

df_prods_clean.shape

(49677, 5)

In [21]:
# Check again for mixed-type data after filtering out missing values

for col in df_prods_clean.columns.tolist():
  weird = (df_prods_clean[[col]].applymap(type) != df_prods_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean[weird]) > 0:
    print (col)

In [ ]:
# Another way you can drop all missing values is via the following command:
# df_prods.dropna(inplace = True)

# If you wanted to use this command to drop only the NaNs from a particular column, the code would look like this:
# df_prods.dropna(subset = [‘product_name’], inplace = True)

# In both cases, rather than creating an entirely new dataframe,
# you’re overwriting df_prods with a new version of df_prods that doesn’t contain the missing values.
# This is done by way of the inplace = True function, which overwrites the original dataframe.
# If you don’t specify an inplace argument in your code, the function will take the default setting, which is inplace = False.
# When specified as False, the command will only return a view of the changed dataframe, leaving the original dataframe untouched.

# As mentioned before, overwriting can be risky.
# Unless you’re absolutely sure it’s safe to drop the values in question, you should create a new dataframe instead.

### Duplicates

In [22]:
# Finding full duplicates within dataframe:

df_dups = df_prods_clean[df_prods_clean.duplicated()]

# This code creates a new subset of df_prods_clean: df_dups, containing only rows that are duplicates.
# The duplicated() function is what identifies duplicate rows.
# It’s run on the df_prods_clean dataframe. Any duplicate rows that it finds are saved within the new df_dups dataframe.

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [ ]:
# df.drop_duplicates()

In [24]:
# Before doing so, let’s check the current number of rows in your df_prods_clean dataframe so that you can compare the number after removing the duplicates

df_prods_clean.shape

(49677, 5)

In [25]:
# Creating a new dataframe that doesn’t include the duplicates
# df_prods_clean_no_dups that contains only the unique rows from df_prods_clean.

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
# You now have 49,672 rows in your dataframe. The five duplicates have been successfully deleted!

df_prods_clean_no_dups.shape

(49672, 5)

### Exercise 4.5 Data Consistency Checks

### Step 1

In [ ]:
# If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

### Step 2 (Descriptive analysis)

In [27]:
# Run the df.describe() function on your df_ords dataframe
# Tip: Keep an eye on min and max values!

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [28]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max value of "prices" seems irrelevant, it's even too high compared to the 3rd quartile (75%)

In [29]:
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### Step 3 (Mixed-type data)

In [32]:
# Check for mixed-type data in your df_ords dataframe.

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


In [33]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### Step 4

In [ ]:
# If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### No mixed-data found

### Step 5 (Missing values)

In [34]:
# Run a check for missing values in your df_ords dataframe.

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [37]:
# Drop unnamed column for df_ords

df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [39]:
# Create subset for the missing values in df_ords

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [40]:
# View missing values subset

df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [41]:
df_ords[df_ords['user_id'] == 1]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [42]:
df_ords[df_ords['user_id'] == 2]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   

    days_since_prior_order  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0

#### Through the two examples above I can tell that any user who has an order number of "1" will have "NaN" as their days_since_prior_order. This make sense since it was their first order, so they wouldn't have any prior orders and that's why they won't have any data showing how long it has been since the previous order. Maximum user id is 206209 and total missing values for days_since_prior_order is also 206209, meaning each user will have one NaN value located in their first order.

### Step 6 (Addressing missing values)

#### I decided not to make any changes. Even though it looks missing, actually the data itself has value in it. We can use it to show that it's actually their first order they've ever made or they're a new customer who just made their first order.

### Step 7 (Duplicates)

In [43]:
# Run a check for duplicate values in your df_ords data.

df_ords_dups = df_ords[df_ords.duplicated()]

In [44]:
# View df_ords_dups

df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [47]:
df_ords_dups.size

0

#### No duplicates found

### Step 8 (Addressing duplicates)

#### No duplicates found

### Step 9 (Export dataframe)

In [48]:
# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [49]:
# Export final df_ords 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))